geo: implement ST_Summary for Geography/Geometry

hueypark opened the pull request:
Fixes #48405, #49049

Release note (sql change): Implemented the geometry based builtins `ST_Summary`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49738)
<!-- Reviewable:end -->

otan commented on docs/generated/sql/functions.md:
```diff
@@ -1051,6 +1051,22 @@ has no relationship with the commit order of concurrent transactions.</p>
 </span></td></tr>
+<tr><td><a name="st_summary"></a><code>st_summary(geography: geography) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Returns a text summary of the contents of the geography.</p>
+<p>Flags shown square brackets after the geometry type have the following meaning:
+M: has M coordinate
```
the generated text here will not look good in HTML.
if you add `*` in front of these in the info string, these should appear as a HTML list, i.e.
```
* M: has M coordinate
* Z: has Z coordinate
* B: has a cached bounding box
* G: is geography
* S: has spatial reference system
```

hueypark replied:
Done.

otan commented on pkg/geo/summary.go:
```diff
@@ -0,0 +1,170 @@
+// S: has spatial reference system
+func Summary(t geom.T, isGeography bool, offset int) (summary string, err error) {
+	shape, err := geopbShape(t)
+	if err != nil {
```
nit: we could pass in `shape` from `geometry.Shape()` or `geography.Shape()` and avoid rechecking this again.

hueypark replied:
Done.

otan commented on pkg/geo/summary.go:
```diff
@@ -0,0 +1,170 @@
+		return "", err
+	}
+
+	f, err := flag(t, isGeography)
```
nit: flag name conflicts with package `flag` and is a little vague. maybe `summaryFlag`?

hueypark replied:
Done.

otan commented on pkg/geo/summary.go:
```diff
@@ -0,0 +1,170 @@
+		numLinearRings := t.NumLinearRings()
+
+		summary += fmt.Sprintf("%s[%s] with %d ring", shape.String(), f, t.NumLinearRings())
+		if 1 < numLinearRings {
```
nit: easier to read this (and below) as `numLinearRings > 1`

hueypark replied:
Done.

otan commented on pkg/geo/summary.go:
```diff
@@ -0,0 +1,170 @@
+
+		for i := 0; i < numPoints; i++ {
+			point := t.Point(i)
+			tmp, err := Summary(point, isGeography, offset+2)
```
nit: the variable named `tmp` here may be better named as `line`

hueypark replied:
Done.

otan commented on pkg/geo/summary.go:
```diff
@@ -0,0 +1,170 @@
+// B: has a cached bounding box
+// G: is geography
+// S: has spatial reference system
+func Summary(t geom.T, isGeography bool, offset int) (summary string, err error) {
```
nit: taking `offset` as an initial argument for the public facing function makes the interface a little unfriendly as there is a dangling magic constant `0`.

Can we do something like:
```
func Summary(t geom.T, isGeography bool) (summary string, err error) {
   return summary(t, isGeography, 0)
}

func summary(t geom.T, isGeography bool, spaceOffset int) (summary string, err error) {
}
```
```

hueypark replied:
Done.

otan reviewed (commented):
thanks for coming again and doing this!
few minor things :)
nit: title can just be `geo: implement ST_Summary for Geography/Geometry`

hueypark committed:
geo: implement ST_Summary for Geography/Geometry

Fixes #48405, #49049

Release note (sql change): Implemented the geometry based builtins `ST_Summary`.

hueypark commented:
Thank you for the review!

> nit: title can just be `geo: implement ST_Summary for Geography/Geometry`

Done.

otan commented on pkg/geo/summary.go:
```diff
@@ -0,0 +1,174 @@
+
+		return sum, nil
+	default:
+		return "", errors.Newf("unspport geom type: %T", t)
```
i've fixed this typo

otan commented on pkg/geo/summary.go:
```diff
@@ -0,0 +1,174 @@
+	return summary(t, shape, isGeography, 0)
+}
+
+func summary(t geom.T, shape geopb.Shape, isGeography bool, offset int) (sum string, err error) {
```
renamed `summaryLine`

otan commented:
bors r+
